Login: Fix user not redirected to initial page when logged out

stephanethomas opened the pull request:
This pull request fixes https://github.com/Automattic/wp-calypso/issues/20941 by addressing the case where users are logged out and try to access a page that requires them to be logged in. In that case they are redirected to the `Login` page, with the url of the previous page added to the `redirect_to` parameter. This parameter will now be honored upon successful login. Finally we should still support the case where users switch to magic link in order to log in to WooCommerce (or any OAuth client app) and must still be redirected to the initial page requested.

The solution chosen here is to use the latest `redirect_to` parameter provided in the url of any login page.

#### Testing instructions

1. Run `git checkout fix/login-redirect` and start your server, or open a [live branch](https://calypso.live/?branch=fix/login-redirect)
2. Follow additional instructions from 1caed-pb to set up your sandbox
 
##### Login with automatic redirect
 
3. Open the [`Jetpack Plans` page](http://calypso.localhost:3000/jetpack/connect/plans/nasty-new-301.jurassic.ninja) for a specific site in an incognito window
4. Assert that you are redirected to the [`Login` page](http://calypso.localhost:3000/log-in?redirect_to=%2Fjetpack%2Fconnect%2Fplans%2Fnasty-new-301.jurassic.ninja)
5. Submit credentials for a regular account
6. Assert that you are logged in, and redirected to the [`Jetpack Plans` page](http://calypso.localhost:3000/jetpack/connect/store)

##### Login with no initial redirect

3. Open the [`Me` page](http://calypso.localhost:3000/me) in an incognito window
4. Click the `Log In` link in the upper right corner of the page
5. Submit credentials for an account with 2FA enabled
6. Assert that you are logged in, and redirected to the [`Me` page](http://calypso.localhost:3000/me) on `wpcalypso`

##### Login with redirect in url

3. Open the [`Login` page](http://calypso.localhost:3000/log-in?redirect_to=http%3A%2F%2Fcalypso.localhost%3A3000%2Fme) with a `redirect_to` parameter in an incognito window
4. Enter the email address or username of a passwordless account, and submit the form
5. Click the link in the email you should have received by now
6. Click the `Finish Login` button
7. Assert that you are logged in, and redirected to the [`Me` page](http://calypso.localhost:3000/me) on `wpcalypso`

##### Magic login with redirect in url

3. Open the [`Magic Login` page](http://calypso.localhost:3000/log-in/link?redirect_to=http%3A%2F%2Fcalypso.localhost%3A3000%2Fstats%2Fday) with a `redirect_to` parameter in an incognito window
4. Enter the email address or username of a regular account, and submit the form
5. Click the link in the email you should have received by now
6. Click the `Finish Login` button
7. Assert that you are logged in, and redirected to the [`Stats` page](http://calypso.localhost:3000/stats/day)

##### WooCommerce with magic login

3. Open the [`WooCommerce Home` page](https://woocommerce.com/) in an incognito window
4. Click the `SIGN IN WITH WORDPRESS.COM` link in the header
5. Assert that you are presented with the `WooCommerce OAuth Login` page
6. Replace https://wordpress.com with http://calypso.localhost:3000 in the address bar, and reload
7. Click the `Email me a login link` link in the footer
8. Enter an email address or a username, and submit the form
9. Click the link in the email you should have received by now
10. Click the `Finish Login` button
11. Assert that you are redirected to the [`WooCommerce OAuth Authorize` page](https://public-api.wordpress.com/oauth2/authorize?client_id=50916)
12. Assert that you are successfully logged in to WordPress.com

##### Jetpack connect

3. [Generate](http://poopy.life/create?src=inquisitive-macaw&key=g57uqr8kfjjn0Pwi) a new Jetpack site in an incognito window
4. Click the `all done! continue to your install` link
5. Click the `Connect to WordPress.com` button in the administration panel
6. Submit credentials for a passwordless account that has 2FA enabled
7. Click the link in the email you should have received by now
8. Click the `Finish Login` button
9. Click the `Approve ` button on the [`Jetpack Connect Authorize` page](https://wordpress.com/jetpack/connect/authorize)
10. Click the `Approve ` button again [if the `Jetpack Connect Authorize` page reloaded](https://github.com/Automattic/wp-calypso/issues/21063)
11. Assert that you are redirected to the [`Jetpack Plans` page](https://wordpress.com/jetpack/connect/plans)
12. Assert a `Your site is now connected!` message is displayed

You can repeat those steps with an account that has 2FA enabled.

#### Reviews
 
- [x] Code
- [x] Product

Tug commented on client/state/selectors/get-redirect-to-from-query-arguments.js:
```diff
@@ -0,0 +1,25 @@
+ * @param {Object} state - global redux state
+ * @return {String?} the redirect_to url
+ */
+export const getRedirectToFromQueryArguments = state => {
```
Maybe
```
export const getLastQueryArguments = state => Object.assign( {}, getInitialQueryArguments( state ), getCurrentQueryArguments( state ) );
```
would be a more generic method

Tug commented:
I tested all the flows mentioned as well as other jetpack flows (such as sso) and I think I found a regression:

When linking my wordpress.com account with a jetpack site it seems to fails during authorization, reloads the page, here I can successfully authorize and I am finally redirected to the

`http://calypso.localhost:3000/jetpack/connect/plans/myjetpacksite.com?redirect=https%3A%2F%2Fmyjetpacksite.com%2Fwp-admin%2Fadmin.php%3Fpage%3Djetpack`

page with the message `You don't have any sites yet.`, instead of being redirected to the jetpack site as expected.

Here is a recording that shows the bug: https://www.useloom.com/share/e99007b8d02943c99ae5574c34daf653

And obviously there is the "authorization on first try" bug which I thought we fixed with the nonce hook on the server. I guess we need to have a look at that as well.

I'd like to work adding more cases to e2e tests for login (and ssr pages in general) so I guess it would be better if we could wait before merging this.

Tug commented:
Just gave it a little more thought, I think we should always use the `redirect_to` from the initial query parameters and make sure any redirect to `/log-in` is a hard redirect:
```js
window.location = login( { native:true, redirectTo } );
```
instead of:
```js
page.redirect( login( { native:true, redirectTo } ) );
```

yurynix commented:
> I'd like to work adding more cases to e2e tests for login (and ssr pages in general) so I guess it would be better if we could wait before merging this.

That's a good suggestion, and maybe we should add more e2e tests, but that's not something we can do overnight ( or am I wrong here? )

We currently have a problem with jetpack sites ( #20941 ), IMO we should fix that ASAP,
so I don't think we should postpone it till we'll get e2e tests cover all login flow.


>  I think we should always use the redirect_to from the initial query parameters and make sure any redirect to /log-in is a hard redirect:

So you mean get rid of all the initial / current query arguments thing and just do "hard redirects"?

seear commented:
> We currently have a problem with jetpack sites ( #20941 ), IMO we should fix that ASAP

The `/jetpack/connect/plans/:site` route login redirect ( #20941) is not an urgent priority because it is not really in use yet. However, any route using [this login lib](https://github.com/Automattic/wp-calypso/blob/master/client/lib/paths/login/index.js) must be affected in the same way?

stephanethomas commented:
> I think we should always use the redirect_to from the initial query parameters and make sure any redirect to /log-in is a hard redirect:

I'm not entirely convinced which should go that route. It's going to hit performance because of the overhead of serving the `Login` page from the server. What about simply better storing this `redirect_to` parameter in the global state tree as soon as it appears in the url (think `ROUTE_SET` event), and then introduce a `redirect_to_sanitized` prop that will be used upon successful login once it has been validated?

> However, any route using this login lib must be affected in the same way?

Indeed.

Tug commented:
> What about simply better storing this redirect_to parameter in the global state tree as soon as it appears in the url

I think it introduces too much complexity.
User might leave the app at any moment and be back to the same page from another flow, we don't want to persist `redirect_to` in any way (except in some special cases like the google connect in redirect mode). Then, even if we make sure we don't persist it there might be situations where a user stays on the app but does not authenticate immediately.

Loading `/log-in?redirect_to=...` from scratch is the safest way we have to make sure users will be redirected to the right url even if they reload the page at some point, simply because it's self contained in the url.

yurynix commented:
> I think it introduces too much complexity.
I agree here, it's already quite complicated, I would love to simplify it and now complicate it more.

stephanethomas commented:
> And obviously there is the "authorization on first try" bug which I thought we fixed with the nonce hook on the server. I guess we need to have a look at that as well.

I've created https://github.com/Automattic/wp-calypso/issues/21063 to address that particular issue. Since that's not a regression introduced by this pull request I would rather fix it in a different one.

stephanethomas commented:
I've rebased, and pushed a proposal for what I had in mind. As you can see, this doesn't introduce more complexity. Putting aide the new reducer and selector, this pull request just updates selector calls.

Finally I've updated the testing instructions with more use cases.

johnHackworth commented:
I've tested the jetpack-related use cases and it works fine and fixes current problems! :+1:

Tug commented:
Indeed it seems to resolve the issue quite nicely. Well done @stephanethomas !
🚢

seear commented:
> I've tested the jetpack-related use cases and it works fine and fixes current problems!

Can confirm this works great for me. Nice work 👍

stephanethomas commented:
Thanks for the review everyone!
